### DISCUSSION TITLE
tutorials/intro-dgs/11-wrapping-up

### DISCUSSION THREAD

**Discussion by giscus** [apollographql-education/content-feedback#83] [course feedback]
# tutorials/intro-dgs/11-wrapping-up

https://www.apollographql.com/tutorials/intro-dgs/11-wrapping-up

**Comment by yamalameyooooo** [comment]
add dgs codegen tutotial along with Connection type

**Reply by mabuyo** [reply]
Hey there @yamalameyooooo - are you referring to the [Relay pagination](https://netflix.github.io/dgs/advanced/relay-pagination/#set-up) `@connection` directive?

**Reply by yamalameyooooo** [reply]
@mabuyo yes, actually i am using the same stack and trying to merge it with apollo federated supergraph, but the supergraph compose fails as PageInfo generated by codegen for Relay Pagination having @connection directive does not have @sharable directive as dgs codegen makes those classes in runtime so the end resulted SDL doesnt have those custom directives which leads to manual declaring the conection types and to write and manage the relay pagination logic in all the java+dgs based subgraphs

**Reply by yamalameyooooo** [reply]
i have opened up a feature request to do so at https://github.com/Netflix/dgs-framework/issues/1762
meanwhile i am just declaring all the connection related types like edge. node, cursor, pageinfo and managing them in the service logic, but this can be done easily using codegen just one drawback is that it wont have @sharable directive or any apollo supergraph based directive, 
if the subgraph is standalone codegen works fine but the moment we try to federate it, there issues pop up

**Reply by lizhennessy** [reply]
Hey there @yamalameyooooo - thanks for pointing this out, and for recording the issue on the DGS Framework project! I'm less familiar with the conventions of Relay Pagination, so forgive my questions: are you starting by writing classes or by writing SDL? (Or perhaps is it a bit of a mix of both?) It does look like the [DGS docs](https://netflix.github.io/dgs/advanced/custom-directives/) describe a couple of different methods of getting custom directives applied, so I want to make sure that I understand this problem thoroughly! Thanks so much for any information or code snippets you can provide.

**Reply by yamalameyooooo** [reply]
@lizhennessy i follow the schema approach first, and sure we can have a demo project which gives the the outline of the problem, 
i did made the same mistake by creating the java based subgraph as per the dgs, 
the problem happened when i created the whole project using the codegen and let the @connection directive handle the relay pagination and its logic (codegen generates the relay based types and classes and logic in runtime, even adds them in SDL)
But the moment we went on to integrate with the supergraph there things started to break as PageInfo specifically was being used by many subgraphs so we are still trying to figure out how to add @sharable directive to those types which are generated in runtime.

Meanwhile we had to define all the Connection types, Edges Types, Nodes, Cursor, PageInfo manually and write and maintain the relay pagination login in the service impl which is very very repetative, 

One of our team mates working in nodejs faced similar issue and he created his own relay based pagination to use at https://github.com/lkrzyzanek/graphql-pagination
but for java we are still trying for dgs team to get an eye on it and add some way where we configure the codegen in "typemapping" and pass the directives we want to define during runtime.

We can also connect over some secure channels to show you the actual issue or blocker which comes up while going for federation

**Comment by mkatsman** [comment]
Should it have the return type? public void getPayloadPlaylist

**Comment by mkatsman** [comment]
At the end, please mention, that it should go to the Apollo sandbox: mutation AddTracksToPlaylist($input: AddItemsToPlaylistInput!)

**Comment by mkatsman** [comment]
Thank you very much the Apollo team. The best tutorial, I ever had. If you need any free help with anything I would be very glad to do  that.
